# E/22/423

# **Foundations of Statistical Inference & Hypothesis Testing**

## **Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)**

1. Derivation of Bias in the Maximum Likelihood Covariance EstimatorWe begin with the Maximum Likelihood Estimator (MLE) for the population covariance matrix $\boldsymbol{\Sigma}$:$$\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$where the empirical sample mean is defined as $\widehat{\boldsymbol{\mu}}_n = \frac{1}{n}\sum_{i=1}^n \mathbf{X}_i$.To evaluate the expectation of this estimator, we decompose the sample deviations by centering them around the true, unobserved population mean $\boldsymbol{\mu}$:$$\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n = (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$$Substituting this expression back into the inner summation of the MLE formula yields:$$\sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T = \sum_{i=1}^n \left[ (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \left[ (\mathbf{X}_i - \boldsymbol{\mu}) - (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right]^T$$Expanding the outer product term-by-term inside the summation gives:$$= \sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T - \sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T - \sum_{i=1}^n (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T + \sum_{i=1}^n (\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$We can simplify the cross-product terms by noting that $\sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu}) = n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$. Because the term $(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})$ does not depend on the index $i$, the summation collapses as follows:$$= \sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T + n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$$$= \sum_{i=1}^n (\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T - n(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T$$Next, we apply the expectation operator $\mathbb{E}[\cdot]$ to the entire expression:$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} \left( \sum_{i=1}^n \mathbb{E}[(\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T] - n \mathbb{E}[(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T] \right)$$By definition of the underlying population parameters, we know that:$\mathbb{E}[(\mathbf{X}_i - \boldsymbol{\mu})(\mathbf{X}_i - \boldsymbol{\mu})^T] = \boldsymbol{\Sigma}$$\mathbb{E}[(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu})^T] = \text{Var}(\widehat{\boldsymbol{\mu}}_n) = \frac{1}{n}\boldsymbol{\Sigma}$Substituting these statistical identities back into our expectation formula:$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{1}{n} \left( n\boldsymbol{\Sigma} - n\left(\frac{1}{n}\boldsymbol{\Sigma}\right) \right) = \frac{1}{n} (n\boldsymbol{\Sigma} - \boldsymbol{\Sigma}) = \left(\frac{n-1}{n}\right)\boldsymbol{\Sigma}$$Conclusion: Because $\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] \neq \boldsymbol{\Sigma}$, the maximum likelihood estimator introduces a systematic downward bias. This mathematical compression happens because the sample data is evaluated relative to its own empirical mean $\widehat{\boldsymbol{\mu}}_n$ rather than the true population mean $\boldsymbol{\mu}$, artificially minimizing the observed variance.Eliminating Bias via Bessel's CorrectionTo construct an unbiased sample covariance estimator ($\mathbf{S}$), we scale the MLE matrix by the reciprocal of the bias factor, $\frac{n}{n-1}$:$$\mathbf{S} = \frac{n}{n-1} \widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{n}{n-1} \left( \frac{1}{n} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T \right) = \frac{1}{n-1} \sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$Verifying the expectation confirms that the bias has been successfully corrected:$$\mathbb{E}[\mathbf{S}] = \frac{n}{n-1} \mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \frac{n}{n-1} \left(\frac{n-1}{n}\right)\boldsymbol{\Sigma} = \boldsymbol{\Sigma}$$2. Structural Health Monitoring: Operational DiagnosticsError FrameworkType I Error ($\alpha$): False Discovery / False Alarm. This occurs when the diagnostic framework incorrectly rejects the null hypothesis ($H_0$). In practice, the system flags a healthy, nominal structure as compromised or damaged, leading to unnecessary downtime or unwarranted maintenance.Type II Error ($\beta$): Missed Target / Unveiled Defect. This occurs when the framework fails to reject a false null hypothesis ($H_0$). Operationally, a structural defect, crack, or failure mode is actively developing, but the diagnostic system fails to trigger an alert, posing severe safety risks.Consequences of an Ultra-Conservative Threshold ($\alpha = 0.0001$)Enforcing a highly restrictive significance level shifts the critical decision boundary far into the extreme tails of the reference distribution. This trade-off induces two major structural consequences:1. Collapse of Statistical Power ($1-\beta$)By making the threshold for anomaly detection exceptionally strict, the system becomes highly resistant to registering any changes. While this virtually eliminates false alarms, it means minor, creeping, or early-stage structural degradation will fail to cross the extreme activation threshold. Consequently, the rate of missed detections ($\beta$) surges, causing the statistical power ($1-\beta$) of the diagnostic system to drastically drop.2. Inflation of the Confidence EllipsoidThe confidence level governing the multi-dimensional acceptance region expands to $1-\alpha = 0.9999$. Geometrically, the volume of the healthy operational confidence ellipsoid expands dramatically.Because the safety boundary swells to encompass nearly all possible data dispersion, the system's sensitivity is compromised, making it largely blind to real-world baseline deviations and structural degradation.

## **Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)**



## 3. Slutsky's Theorem

Let $\{X_n\}$ and $\{Y_n\}$ be sequences of random variables (or random vectors). If $X_n$ converges in distribution to a random variable $X$ ($X_n \xrightarrow{d} X$), and $Y_n$ converges in probability to a constant scalar or matrix $c$ ($Y_n \xrightarrow{p} c$), then the following asymptotic properties hold:

* $X_n + Y_n \xrightarrow{d} X + c$
* $X_n Y_n \xrightarrow{d} cX$
* $X_n / Y_n \xrightarrow{d} X/c$ — *provided that $c \neq 0$ (or $\mathbf{c}$ is nonsingular in matrix contexts).*

---

## 4. Asymptotic Distribution Justification

### Classical Multivariate Central Limit Theorem

By the Lindeberg-Lévy Multivariate CLT, the centered and scaled sample mean vector asymptotically approaches a multivariate normal distribution:

$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$

By applying a linear transformation using the inverse matrix square root of the true population covariance matrix $\boldsymbol{\Sigma}^{-1/2}$, we map this space into a standard, uncorrelated identity frame:

$$\boldsymbol{\Sigma}^{-1/2} \left[ \sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \xrightarrow{d} \mathscr{N}(\mathbf{0}, \mathbf{I})$$

### Convergence of the Sample Covariance

According to the Weak Law of Large Numbers (WLLN), the unbiased sample covariance matrix $\mathbf{S}$ serves as a consistent estimator of the population covariance $\boldsymbol{\Sigma}$. Consequently, as the sample size $n$ approaches infinity, $\mathbf{S}$ converges in probability to the deterministic true matrix:

$$\mathbf{S} \xrightarrow{p} \boldsymbol{\Sigma}$$

Because the inverse matrix square root function is continuous over the space of positive-definite matrices, Continuous Mapping Theorem ensures that:

$$\mathbf{S}^{-1/2} \xrightarrow{p} \boldsymbol{\Sigma}^{-1/2}$$

### Synthesis via Slutsky's Theorem

We now invoke Slutsky’s Theorem to couple our two converging sequences. By pre-multiplying the sequence that converges in distribution by the sequence that converges in probability to a constant matrix, we establish that:

$$\mathbf{S}^{-1/2} \left[ \sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \right] \xrightarrow{d} \boldsymbol{\Sigma}^{-1/2} \cdot \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma}) = \mathscr{N}(\mathbf{0}, \mathbf{I})$$

> **Operational Significance:** Because the resulting target distribution scales cleanly to a standard normal distribution, we can practically invert this transformation using the empirical sample matrix instead of the unknown population parameter. For large sample sizes $n$, this yields the working approximation:
> $$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \sim \mathscr{N}(\mathbf{0}, \mathbf{S}) \implies \widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\left(\boldsymbol{\mu}, \frac{1}{n}\mathbf{S}\right)$$
>
>
> This mathematically justifies substituting the sample covariance matrix $\mathbf{S}$ in place of the unknown true covariance $\boldsymbol{\Sigma}$ when constructing large-sample confidence bounds, guaranteeing that the asymptotic integrity of the diagnostic threshold remains stable.

## **Part C: Numerical Verification**

In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

# Generating base stationary multivariate normal noise
base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

# Injecting a subtle, hidden structural drift in the final 1,000 snapshots
base_data[4000:, 0] += 0.015  # Sensor 1 drift
base_data[4000:, 2] -= 0.010  # Sensor 3 drift

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partitions the dataset into g chunks and evaluates first-moment homogeneity
    via Wilks' Lambda and Bartlett's Chi-Square asymptotic transformation.
    Calculated completely from scratch.
    """
    X = df.to_numpy()
    n, m = X.shape
    n_j = n // g_chunks  # Size of each chunk (1000)

    # 1. Compute global grand mean
    grand_mean = np.mean(X, axis=0)

    # Initialize Matrix W and Matrix B
    W = np.zeros((m, m))
    B = np.zeros((m, m))

    # 2. Iterate through consecutive, non-overlapping blocks
    for j in range(g_chunks):
        chunk = X[j * n_j : (j + 1) * n_j]
        chunk_mean = np.mean(chunk, axis=0)

        # Within-Chunk sum of squares and cross-products (W)
        # W_j = sum_{i} (x_ij - mean_j)(x_ij - mean_j)^T
        dev_within = chunk - chunk_mean
        W += dev_within.T @ dev_within

        # Between-Chunk sum of squares and cross-products (B)
        # B_j = n_j * (mean_j - grand_mean)(mean_j - grand_mean)^T
        dev_between = (chunk_mean - grand_mean).reshape(-1, 1)
        B += n_j * (dev_between @ dev_between.T)

    # 3. Calculate Wilks' Lambda: |W| / |B + W|
    det_W = np.linalg.det(W)
    det_T = np.linalg.det(B + W)
    wilks_lambda = det_W / det_T

    # 4. Bartlett's Chi-Square transformation
    adjustment = n - 1 - (m + g_chunks) / 2
    chi2_calc = -adjustment * np.log(wilks_lambda)

    # Degrees of freedom: m * (g - 1)
    df_chi2 = m * (g_chunks - 1)

    # Calculate p-value
    p_value = stats.chi2.sf(chi2_calc, df_chi2)

    return {
        "Wilks_Lambda": float(wilks_lambda),
        "Chi2_Stat": float(chi2_calc),
        "Degrees_of_Freedom": int(df_chi2),
        "p_value": float(p_value)
    }

# Execute your completed function
results = verify_first_moment_homogeneity(df_strain, g_chunks=5)

print("="*60)
print("             MANOVA DIAGNOSTIC REPORT                       ")
print("="*60)
print(f"Calculated Wilks' Lambda (Λ): {results['Wilks_Lambda']:.6f}")
print(f"Bartlett's Chi-Square (χ²):    {results['Chi2_Stat']:.4f}")
print(f"Degrees of Freedom (df):       {results['Degrees_of_Freedom']}")
print(f"Calculated p-value:            {results['p_value']:.6f}")
print("-"*60)

alpha = 0.05
print(f"Significance Threshold (α):    {alpha}")
if results['p_value'] < alpha:
    print("CONCLUSION: REJECT H0.")
    print("Diagnostic: The baseline center has statistically shifted over time.")
    print("            First Moment is HETEROGENEOUS (Structural drift detected).")
else:
    print("CONCLUSION: FAIL TO REJECT H0.")
    print("Diagnostic: First Moment is HOMOGENEOUS. Baseline center is stable.")
print("="*60)

             MANOVA DIAGNOSTIC REPORT                       
Calculated Wilks' Lambda (Λ): 0.990452
Bartlett's Chi-Square (χ²):    47.9215
Degrees of Freedom (df):       12
Calculated p-value:            0.000003
------------------------------------------------------------
Significance Threshold (α):    0.05
CONCLUSION: REJECT H0.
Diagnostic: The baseline center has statistically shifted over time.
            First Moment is HETEROGENEOUS (Structural drift detected).


# **Geometric Subspace Optimization via Principal Component Analysis (PCA)**

## **Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)**

We are given the orthogonal projection mapping $\mathbf{Z}_i = \mathbf{P}^T \widetilde{\mathbf{X}}_i$, where $\widetilde{\mathbf{X}}_i$ is zero-mean ($\mathbb{E}[\widetilde{\mathbf{X}}_i] = \mathbf{0}$), and its population covariance matrix is $\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$. The spectral decomposition of this symmetric, positive-semidefinite matrix is $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$, where $\mathbf{P}$ is an orthogonal matrix ($\mathbf{P}^T\mathbf{P} = \mathbf{P}\mathbf{P}^T = \mathbf{I}$) containing the eigenvectors, and $\mathbf{\Lambda}$ is a diagonal matrix containing the eigenvalues.To find the theoretical covariance matrix of the transformed vector $\mathbf{Z}_i$, we evaluate the expectation of its outer product:$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}\left[ (\mathbf{P}^T \widetilde{\mathbf{X}}_i) (\mathbf{P}^T \widetilde{\mathbf{X}}_i)^T \right]$$Applying the matrix transpose property $(\mathbf{A}\mathbf{B})^T = \mathbf{B}^T\mathbf{A}^T$:$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}\left[ \mathbf{P}^T \widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T \mathbf{P} \right]$$Because expectation is a linear operator and the matrix of eigenvectors $\mathbf{P}$ is deterministic (non-random), we can pull $\mathbf{P}^T$ and $\mathbf{P}$ outside of the expectation operator:$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{P}^T \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T] \mathbf{P}$$Substitute the definition of the baseline population covariance matrix $\boldsymbol{\Sigma} = \mathbb{E}[\widetilde{\mathbf{X}}_i \widetilde{\mathbf{X}}_i^T]$ into the expression:$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P}$$Now, substitute the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P} \mathbf{\Lambda} \mathbf{P}^T$ into the equation:$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{P}^T (\mathbf{P} \mathbf{\Lambda} \mathbf{P}^T) \mathbf{P}$$By the associative property of matrix multiplication:$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = (\mathbf{P}^T \mathbf{P}) \mathbf{\Lambda} (\mathbf{P}^T \mathbf{P})$$Using the property of the orthogonal matrix ($\mathbf{P}^T \mathbf{P} = \mathbf{I}$):$$\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{I} \mathbf{\Lambda} \mathbf{I} = \mathbf{\Lambda}$$Thus, the theoretical covariance matrix of the transformed vector simplifies exactly to the diagonal matrix $\mathbf{\Lambda}$. $\blacksquare$Geometric and Statistical MeaningStatistical Meaning: Because the resulting matrix $\mathbf{\Lambda}$ is strictly diagonal, the off-diagonal elements are exactly zero:$$\mathbb{E}[Z_{i,j} Z_{i,k}] = 0 \quad \forall j \neq k$$Since the transformed features are zero-mean ($\mathbb{E}[\mathbf{Z}_i] = \mathbf{P}^T \mathbb{E}[\widetilde{\mathbf{X}}_i] = \mathbf{0}$), this proves that the cross-covariance—and by extension, the linear correlation—between any two distinct principal components $Z_{i,j}$ and $Z_{i,k}$ is exactly zero. The transformation strips away all redundant linear dependencies from the raw data.Geometric Meaning: The orthogonal matrix $\mathbf{P}^T$ acts as a rigid rotation operator. It rotates the original, highly correlated coordinate axes of $\mathbb{R}^m$ into a new orthogonal coordinate framework. In this new frame, the axes align perfectly with the hyper-ellipsoidal principal directions of data dispersion. The diagonal elements $\lambda_j$ represent the variance (spread) of the data along each of these newly defined, mutually perpendicular geometric axes.Part 2: Variance Conservation & Subspace PartitioningFormal Proof of Variance ConservationThe total variance of a multivariate system in its raw coordinate space is defined as the sum of the variances of its individual features, which corresponds to the trace of its covariance matrix, $\text{tr}(\boldsymbol{\Sigma})$.We evaluate the trace of the diagonalized matrix $\mathbf{\Lambda}$. From the spectral decomposition proof, we know that $\mathbf{\Lambda} = \mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P}$. Taking the trace of both sides:$$\text{tr}(\mathbf{\Lambda}) = \text{tr}(\mathbf{P}^T \boldsymbol{\Sigma} \mathbf{P})$$The cyclic permutation property of the trace operator states that for any matrices $\mathbf{A}$ and $\mathbf{B}$ such that the products are square, $\text{tr}(\mathbf{A}\mathbf{B}) = \text{tr}(\mathbf{B}\mathbf{A})$. Grouping our matrices as $\mathbf{A} = \mathbf{P}^T$ and $\mathbf{B} = \boldsymbol{\Sigma} \mathbf{P}$, we cyclically permute $\mathbf{A}$ to the end of the product:$$\text{tr}(\mathbf{P}^T (\boldsymbol{\Sigma} \mathbf{P})) = \text{tr}((\boldsymbol{\Sigma} \mathbf{P}) \mathbf{P}^T)$$By associativity:$$\text{tr}(\mathbf{\Lambda}) = \text{tr}(\boldsymbol{\Sigma} (\mathbf{P} \mathbf{P}^T))$$Because $\mathbf{P}$ is an orthogonal matrix, its outer product also yields the identity matrix ($\mathbf{P}\mathbf{P}^T = \mathbf{I}$). Substituting this back into our expression:$$\text{tr}(\mathbf{\Lambda}) = \text{tr}(\boldsymbol{\Sigma} \mathbf{I}) = \text{tr}(\boldsymbol{\Sigma})$$Since $\mathbf{\Lambda}$ is a diagonal matrix, its trace is simply the sum of its diagonal entries (the eigenvalues):$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{\Lambda}) = \sum_{j=1}^m \lambda_j$$This proves that the total system variance is preserved during the orthogonal mapping. $\blacksquare$Subspace Partitioning FormulationsWhen partitioning our empirical system using the ordered sample eigenvalues ($\widehat{\lambda}_1 \ge \widehat{\lambda}_2 \ge \dots \ge \widehat{\lambda}_m$), the mathematical definitions for the variance ratios at a given cutoff $k$ are defined as:Cumulative Explained Variance Ratio $\Phi(k)$: Measures the proportion of total systemic energy retained within the primary $k$-dimensional principal subspace.$$\Phi(k) = \frac{\text{tr}(\widehat{\mathbf{P}}_k^T \mathbf{S} \widehat{\mathbf{P}}_k)}{\text{tr}(\mathbf{S})} = \frac{\sum_{j=1}^k \widehat{\lambda}_j}{\sum_{j=1}^m \widehat{\lambda}_j}$$Residual Unexplained Variance Ratio $\Psi(k)$: Measures the remaining proportion of total systemic energy cast off into the $(m-k)$-dimensional residual subspace.$$\Psi(k) = \frac{\text{tr}(\widehat{\mathbf{P}}_{m-k}^T \mathbf{S} \widehat{\mathbf{P}}_{m-k})}{\text{tr}(\mathbf{S})} = \frac{\sum_{j=k+1}^m \widehat{\lambda}_j}{\sum_{j=1}^m \widehat{\lambda}_j} = 1 - \Phi(k)$$Part 3: Subspace Energy Decomposition & DiagnosticsVector DefinitionsGiven the full orthogonal basis expansion $(\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) = \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$:Reconstructed Vector ($\widehat{\mathbf{x}}_i$): The projection of the observation back into the raw data space from the truncated $k$-dimensional principal subspace.$$\widehat{\mathbf{x}}_i = \widehat{\boldsymbol{\mu}}_n + \widehat{\mathbf{P}}_k \mathbf{z}_{i,k}$$Residual Error Vector ($\mathbf{e}_i$): The discrepancy vector representing the information lost due to truncation, which lies entirely within the residual subspace.$$\mathbf{e}_i = (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) - (\widehat{\mathbf{x}}_i - \widehat{\boldsymbol{\mu}}_n) = \mathbf{x}_i - \widehat{\mathbf{x}}_i = \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}$$Pythagorean Energy Decomposition ProofWe express the squared Euclidean norm of the centered observation vector as an inner product:$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)^T (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n)$$Substituting the partitioned subspace representation into the inner product:$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \left( \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k} \right)^T \left( \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k} \right)$$Distributing the transpose operator using $(\mathbf{A}+\mathbf{B})^T = \mathbf{A}^T + \mathbf{B}^T$ and expanding the terms:$$\begin{aligned}
\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \ & \mathbf{z}_{i,k}^T \widehat{\mathbf{P}}_k^T \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \mathbf{z}_{i,k}^T \widehat{\mathbf{P}}_k^T \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k} \\
& + \mathbf{z}_{i,m-k}^T \widehat{\mathbf{P}}_{m-k}^T \widehat{\mathbf{P}}_k \mathbf{z}_{i,k} + \mathbf{z}_{i,m-k}^T \widehat{\mathbf{P}}_{m-k}^T \widehat{\mathbf{P}}_{m-k} \mathbf{z}_{i,m-k}
\end{aligned}$$Because the full eigenvector matrix $\widehat{\mathbf{P}} = [\widehat{\mathbf{P}}_k \ \widehat{\mathbf{P}}_{m-k}]$ is orthogonal, its sub-bases satisfy these properties:Subspace Orthonormal Normalization: $\widehat{\mathbf{P}}_k^T \widehat{\mathbf{P}}_k = \mathbf{I}_{k \times k}$ and $\widehat{\mathbf{P}}_{m-k}^T \widehat{\mathbf{P}}_{m-k} = \mathbf{I}_{(m-k) \times (m-k)}$Mutual Subspace Orthogonality: $\widehat{\mathbf{P}}_k^T \widehat{\mathbf{P}}_{m-k} = \mathbf{0}_{k \times (m-k)}$ and $\widehat{\mathbf{P}}_{m-k}^T \widehat{\mathbf{P}}_k = \mathbf{0}_{(m-k) \times k}$Applying these properties simplifies the expanded equation by eliminating the cross-product terms:$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \mathbf{z}_{i,k}^T \mathbf{I}_{k \times k} \mathbf{z}_{i,k} + \mathbf{0} + \mathbf{0} + \mathbf{z}_{i,m-k}^T \mathbf{I}_{(m-k) \times (m-k)} \mathbf{z}_{i,m-k}$$$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \mathbf{z}_{i,k}^T \mathbf{z}_{i,k} + \mathbf{z}_{i,m-k}^T \mathbf{z}_{i,m-k}$$Using the definition of the squared Euclidean norm ($\|\mathbf{v}\|^2 = \mathbf{v}^T\mathbf{v}$):$$\|\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n\|^2 = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2$$Thus, the total energy decomposes cleanly into independent subspace components. $\blacksquare$Diagnostic Function Contrast & Structural MappingHotelling's $T^2$ Statistic: Measures the Mahalanobis distance within the principal subspace ($\widehat{\mathbf{P}}_k$). It monitors variations within the established, high-energy operational boundaries of the normal design space.Residual $Q$ Statistic (SPE): Measures the squared Euclidean norm of the residual vector within the unexplained subspace ($\widehat{\mathbf{P}}_{m-k}$). It monitors variations that break the established physical correlations, manifesting as energy emerging outside the normal design space.Event Mapping & JustificationEnvironmental Load Anomaly (Extreme Operational Wind Loads) $\longrightarrow$ Hotelling's $T^2$ StatisticJustification: Extreme wind loads alter the operational amplitude of structural vibrations but do not break the underlying physical properties of the structure. The sensors record larger measurements along known, correlated operational pathways. This inflation of energy within the primary subspace drives up the Mahalanobis distance, making Hotelling's $T^2$ the most sensitive detector.Internal Structural Fracture (Fatigue Crack) $\longrightarrow$ Residual $Q$ Statistic (SPE)Justification: A physical crack permanently changes the structural geometry and load distribution pathways. This alters the underlying boundary conditions, breaking the cross-channel correlations established in the baseline matrix $\mathbf{S}$. The resulting new dynamic behavior cannot be mapped by the baseline eigenvectors $\widehat{\mathbf{P}}_k$, causing a major leakage of energy into the residual subspace. This triggers a sharp spike in the $Q$ statistic.

In [4]:
import numpy as np
import pandas as pd

class SHMDiagnosticsPipeline:
    def __init__(self, n_samples: int = 3000, m_sensors: int = 4):
        self.n = n_samples
        self.m = m_sensors
        self.df_baseline = self._generate_synthetic_baseline()

    def _generate_synthetic_baseline(self) -> pd.DataFrame:
        """Generates representative multi-channel nominal SHM baseline data."""
        np.random.seed(2026)
        # Defining baseline mean operational strains
        mean_vector = [15.2, 22.8, -8.5, 4.1]
        # Strong cross-channel correlations due to structural connectivity
        cov_matrix = [
            [0.25, 0.18, -0.10, 0.05],
            [0.18, 0.36, -0.12, 0.08],
            [-0.10, -0.12, 0.16, -0.04],
            [0.05, 0.08, -0.04, 0.09]
        ]
        raw_data = np.random.multivariate_normal(mean_vector, cov_matrix, size=self.n)
        columns = [f'Sensor_Ch{i+1}' for i in range(self.m)]
        return pd.DataFrame(raw_data, columns=columns)

    def analyze_truncation_boundaries(self) -> dict:
        """
        Standardizes data, computes the spectral decomposition from scratch,
        and tracks Hotelling T2 and Q statistics across k cutoffs.
        """
        # 1. Extract numpy array and center data
        X = self.df_baseline.to_numpy()
        sample_mean = np.mean(X, axis=0)
        X_centered = X - sample_mean

        # 2. Compute unbiased sample covariance matrix S
        S = (X_centered.T @ X_centered) / (self.n - 1)

        # Eigen-decomposition
        eigenvalues, eigenvectors = np.linalg.eigh(S)

        # Sort in descending order
        idx = np.argsort(eigenvalues)[::-1]
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]

        # Project full dataset into principal space (Z = X_centered * P)
        # Shape: (n, m)
        Z = X_centered @ eigenvectors

        summary_metrics = {}

        # 3. Loop through possible subspace cutoffs k
        for k in [1, 2, 3]:
            # Hotelling's T^2 calculation for each sample row
            # T^2 = sum_{j=1}^k (z_j^2 / lambda_j)
            Z_k = Z[:, :k]
            lambdas_k = eigenvalues[:k]
            t2_samples = np.sum((Z_k ** 2) / lambdas_k, axis=1)

            # Residual Q statistic calculation for each sample row
            # Q = sum_{j=k+1}^m z_j^2
            Z_residual = Z[:, k:]
            q_samples = np.sum(Z_residual ** 2, axis=1)

            # 4. Compute expected empirical values
            mean_t2 = np.mean(t2_samples)
            mean_q = np.mean(q_samples)

            # Theoretical expected values for verification:
            # E[T^2] = k,  E[Q] = sum_{j=k+1}^m ((n-1)/n) * lambda_j
            theoretical_t2 = float(k)
            theoretical_q = float(np.sum(eigenvalues[k:]) * ((self.n - 1) / self.n))

            summary_metrics[k] = {
                "Observed_Mean_T2": float(mean_t2),
                "Theoretical_Expectation_T2": theoretical_t2,
                "Observed_Mean_Q": float(mean_q),
                "Theoretical_Expectation_Q": theoretical_q,
                "Subspace_Eigenvalues": eigenvalues.tolist()
            }

        return summary_metrics

# =====================================================================
# Pipeline Execution & Verification
# =====================================================================
pipeline = SHMDiagnosticsPipeline()
analysis = pipeline.analyze_truncation_boundaries()

print("="*75)
print("             SHM SUBSPACE TRUNCATION ANALYSIS REPORT                   ")
print("="*75)
for k, metrics in analysis.items():
    print(f"Subspace Cutoff Boundary [ k = {k} ]")
    print(f"  -> Hotelling's T² Subspace:")
    print(f"     Empirical Mean T²: {metrics['Observed_Mean_T2']:.6f} | Theoretical E[T²]: {metrics['Theoretical_Expectation_T2']:.6f}")
    print(f"  -> Residual Q Subspace:")
    print(f"     Empirical Mean Q:  {metrics['Observed_Mean_Q']:.6f} | Theoretical E[Q]:  {metrics['Theoretical_Expectation_Q']:.6f}")
    print("-"*75)

             SHM SUBSPACE TRUNCATION ANALYSIS REPORT                   
Subspace Cutoff Boundary [ k = 1 ]
  -> Hotelling's T² Subspace:
     Empirical Mean T²: 0.999667 | Theoretical E[T²]: 1.000000
  -> Residual Q Subspace:
     Empirical Mean Q:  0.288158 | Theoretical E[Q]:  0.288158
---------------------------------------------------------------------------
Subspace Cutoff Boundary [ k = 2 ]
  -> Hotelling's T² Subspace:
     Empirical Mean T²: 1.999333 | Theoretical E[T²]: 2.000000
  -> Residual Q Subspace:
     Empirical Mean Q:  0.163588 | Theoretical E[Q]:  0.163588
---------------------------------------------------------------------------
Subspace Cutoff Boundary [ k = 3 ]
  -> Hotelling's T² Subspace:
     Empirical Mean T²: 2.999000 | Theoretical E[T²]: 3.000000
  -> Residual Q Subspace:
     Empirical Mean Q:  0.066664 | Theoretical E[Q]:  0.066664
---------------------------------------------------------------------------


# **Latent Subspace Decomposition via Factor Analysis (FA)**

## **Part A: Theoretical Fundamentals (Generative Model & Commonalities)**

We begin with the linear generative Factor Analysis model for a normalized, zero-mean baseline feature vector $\mathbf{Z}_i \in \mathbb{R}^m$:$$\mathbf{Z}_i = \boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i$$where $\boldsymbol{\Lambda} \in \mathbb{R}^{m \times k}$ represents the factor loading matrix, $\mathbf{F}_i \in \mathbb{R}^k$ represents the unobservable latent common factors, and $\boldsymbol{\epsilon}_i \in \mathbb{R}^m$ represents the unique errors (specific factors). We are given the following structural expectations:$\mathbb{E}[\mathbf{Z}_i] = \mathbf{0}$, $\mathbb{E}[\mathbf{F}_i] = \mathbf{0}$, $\mathbb{E}[\boldsymbol{\epsilon}_i] = \mathbf{0}$$\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}_{k \times k}$ (orthogonal common factors)$\mathbb{E}[\boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T] = \boldsymbol{\Psi} = \text{diag}(\varphi_1^2, \dots, \varphi_m^2)$ (uncorrelated unique errors)$\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}_{m \times k}$ (independence between latent factors and errors, which also implies $\mathbb{E}[\mathbf{F}_i \boldsymbol{\epsilon}_i^T] = \mathbf{0}_{k \times m}$)The universal correlation matrix $\mathbf{R}$ is defined as the expected outer product of the feature vector:$$\mathbf{R} = \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbb{E}\left[ (\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)(\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)^T \right]$$Applying the distributive properties of matrix transpositions and multiplications:$$\mathbf{R} = \mathbb{E}\left[ (\boldsymbol{\Lambda} \mathbf{F}_i + \boldsymbol{\epsilon}_i)(\mathbf{F}_i^T \boldsymbol{\Lambda}^T + \boldsymbol{\epsilon}_i^T) \right]$$$$\mathbf{R} = \mathbb{E}\left[ \boldsymbol{\Lambda} \mathbf{F}_i \mathbf{F}_i^T \boldsymbol{\Lambda}^T + \boldsymbol{\Lambda} \mathbf{F}_i \boldsymbol{\epsilon}_i^T + \boldsymbol{\epsilon}_i \mathbf{F}_i^T \boldsymbol{\Lambda}^T + \boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T \right]$$Using the linearity of the expectation operator:$$\mathbf{R} = \boldsymbol{\Lambda} \mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] \boldsymbol{\Lambda}^T + \boldsymbol{\Lambda} \mathbb{E}[\mathbf{F}_i \boldsymbol{\epsilon}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] \boldsymbol{\Lambda}^T + \mathbb{E}[\boldsymbol{\epsilon}_i \boldsymbol{\epsilon}_i^T]$$Substituting the structural expectations ($\mathbf{I}_{k \times k}$, $\mathbf{0}$, and $\boldsymbol{\Psi}$) into this expression:$$\mathbf{R} = \boldsymbol{\Lambda} (\mathbf{I}_{k \times k}) \boldsymbol{\Lambda}^T + \boldsymbol{\Lambda} (\mathbf{0}_{k \times m}) + (\mathbf{0}_{m \times k}) \boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$$$\mathbf{R} = \boldsymbol{\Lambda} \boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$Thus, the global covariance structure partitions cleanly into common and unique components. $\blacksquare$Parameter Definitions and Asset Diagnostics InterpretationsExtracting the $j$-th diagonal entry of the fundamental equation yields:$$R_{jj} = 1 = \sum_{l=1}^k \lambda_{jl}^2 + \varphi_j^2$$Communality ($h_j^2 = \sum_{l=1}^k \lambda_{jl}^2$): The portion of the variance of the $j$-th sensor explained by the $k$ shared latent factors.Diagnostics Interpretation: Communality represents the system-integrated energy. A high communality indicates that the sensor is highly correlated with the broader structural or operational state of the asset (e.g., global thermal loads, systemic vibrations). If an asset shifts its operational mode, these shared factors reflect that change, making high-communality sensors highly effective for monitoring global process changes.Uniqueness ($\varphi_j^2$): The remaining portion of the variance unique to the $j$-th sensor, combining its specific physical localized variance and raw measurement noise.Diagnostics Interpretation: Uniqueness represents the isolated localized energy. In asset diagnostics, a sudden increase in a sensor's uniqueness means its readings are breaking away from the common systemic framework. This localized decoupling is a classic signature of localized faults, such as a single failing transducer, localized loose housing, or an isolated structural crack near that specific sensor.Part 2: Factor Rotation and InterpretabilityConceptual Contrast: PCA vs. Varimax-Rotated LoadingsRaw PCA Loading Vector: Derived purely from a geometric optimization goal: maximize the variance explained sequentially by each independent axis. Consequently, the first principal component acts as a "global factor," where almost all features have moderate-to-high loadings. While mathematically optimal for data compression, these unrotated loadings are often difficult to interpret physically because every sensor maps to every component simultaneously.Varimax-Rotated Factor Loading Vector: Maximizes the variance of the squared loadings within the columns of $\boldsymbol{\Lambda}$. It adjusts the coordinate system to achieve a "simple structure": each feature loads heavily onto only one (or a few) latent factors, and its loadings on the remaining factors drop near zero. Conceptually, this aligns the latent axes with distinct physical sub-phenomena (e.g., separating structural torsion from purely thermal expansion), making the model far more interpretable for engineering diagnostics.Preservation of Communalities and Global CovarianceLet $\mathbf{T} \in \mathbb{R}^{k \times k}$ be an orthogonal rotation matrix satisfying $\mathbf{T}\mathbf{T}^T = \mathbf{T}^T\mathbf{T} = \mathbf{I}_{k \times k}$. The rotated factor loading matrix is defined as $\boldsymbol{\Lambda}^* = \boldsymbol{\Lambda}\mathbf{T}$.1. Global Covariance Approximation InvarianceWe substitute the rotated loading matrix into the fundamental equation structure to verify if the global approximation changes:$$\mathbf{R}^* = \boldsymbol{\Lambda}^*(\boldsymbol{\Lambda}^*)^T + \boldsymbol{\Psi}$$$$\mathbf{R}^* = (\boldsymbol{\Lambda}\mathbf{T})(\boldsymbol{\Lambda}\mathbf{T})^T + \boldsymbol{\Psi}$$Applying the matrix transpose property $(\mathbf{A}\mathbf{B})^T = \mathbf{B}^T\mathbf{A}^T$:$$\mathbf{R}^* = \boldsymbol{\Lambda}(\mathbf{T}\mathbf{T}^T)\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$Since $\mathbf{T}$ is orthogonal, $\mathbf{T}\mathbf{T}^T = \mathbf{I}$:$$\mathbf{R}^* = \boldsymbol{\Lambda}\mathbf{I}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi} = \mathbf{R}$$The global covariance approximation is completely invariant to the rotation.2. Individual Sensor Communality InvarianceThe communality of the $j$-th sensor is the $j$-th diagonal element of the matrix product $\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$, denoted as $[\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T]_{jj}$. Evaluating this value using the rotated loadings:$$h_j^{*2} = [\boldsymbol{\Lambda}^*(\boldsymbol{\Lambda}^*)^T]_{jj} = [\boldsymbol{\Lambda}\mathbf{T}\mathbf{T}^T\boldsymbol{\Lambda}^T]_{jj} = [\boldsymbol{\Lambda}\mathbf{I}\boldsymbol{\Lambda}^T]_{jj} = [\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T]_{jj} = h_j^2$$Because the matrix product simplifies identically due to the property of the orthogonal rotation matrix ($\mathbf{T}\mathbf{T}^T = \mathbf{I}$), individual column entries within a single row of $\boldsymbol{\Lambda}$ can change significantly, but their row sum of squares ($h_j^2$) remains unchanged.Part 3: Latent Factor Estimation via Thomson's RegressionProof of the Joint Partitioned Covariance MatrixLet $\mathbf{Y}_i = \begin{bmatrix} \mathbf{Z}_i \\ \mathbf{F}_i \end{bmatrix} \in \mathbb{R}^{(m+k)}$. Given that both vectors are zero-mean, the joint covariance matrix $\boldsymbol{\Sigma}_{YY}$ is defined by blocks:$$\boldsymbol{\Sigma}_{YY} = \mathbb{E}[\mathbf{Y}_i \mathbf{Y}_i^T] = \mathbb{E}\left[ \begin{bmatrix} \mathbf{Z}_i \\ \mathbf{F}_i \end{bmatrix} \begin{bmatrix} \mathbf{Z}_i^T & \mathbf{F}_i^T \end{bmatrix} \right] = \begin{bmatrix} \mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] & \mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] \\ \mathbb{E}[\mathbf{F}_i \mathbf{Z}_i^T] & \mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] \end{bmatrix}$$We evaluate each block using our model assumptions:Top-Left Block ($\boldsymbol{\Sigma}_{11}$): By definition, $\mathbb{E}[\mathbf{Z}_i \mathbf{Z}_i^T] = \mathbf{R}$.Bottom-Right Block ($\boldsymbol{\Sigma}_{22}$): By assumption, the factors are normalized and orthogonal: $\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}_{k \times k}$.Top-Right Block ($\boldsymbol{\Sigma}_{12}$): Substitute the generative model $\mathbf{Z}_i = \boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i$:$$\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)\mathbf{F}_i^T] = \boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T]$$Substituting $\mathbb{E}[\mathbf{F}_i \mathbf{F}_i^T] = \mathbf{I}_{k \times k}$ and the independence condition $\mathbb{E}[\boldsymbol{\epsilon}_i \mathbf{F}_i^T] = \mathbf{0}$:$$\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T] = \boldsymbol{\Lambda}\mathbf{I}_{k \times k} + \mathbf{0} = \boldsymbol{\Lambda}$$Bottom-Left Block ($\boldsymbol{\Sigma}_{21}$): This is the transpose of the top-right block:$$\mathbb{E}[\mathbf{F}_i \mathbf{Z}_i^T] = (\mathbb{E}[\mathbf{Z}_i \mathbf{F}_i^T])^T = \boldsymbol{\Lambda}^T$$Assembling these blocks yields the joint partitioned covariance matrix:$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k \times k} \end{bmatrix} \quad \blacksquare$$Derivation of Thomson's Score EquationUnder the Multivariate Normal assumption, we apply the conditional mean projection identity to our partitioned system. Let $\mathbf{x}_1 \rightarrow \mathbf{z}_i$ and $\mathbf{X}_2 \rightarrow \mathbf{F}_i$. The respective mean vectors are $\boldsymbol{\mu}_1 = \mathbf{0}$ and $\boldsymbol{\mu}_2 = \mathbf{0}$.From our derived joint covariance matrix $\boldsymbol{\Sigma}_{YY}$, we map the sub-blocks to the conditional mean identity:$\boldsymbol{\Sigma}_{11} = \mathbf{R}$ (hence $\boldsymbol{\Sigma}_{11}^{-1} = \mathbf{R}^{-1}$)$\boldsymbol{\Sigma}_{21} = \boldsymbol{\Lambda}^T$Substitute these sub-blocks directly into the conditional expectation formula:$$\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i | \mathbf{z}_i] = \boldsymbol{\mu}_2 + \boldsymbol{\Sigma}_{21} \boldsymbol{\Sigma}_{11}^{-1} (\mathbf{z}_i - \boldsymbol{\mu}_1)$$$$\mathbf{f}_i = \mathbf{0} + \boldsymbol{\Lambda}^T \mathbf{R}^{-1} (\mathbf{z}_i - \mathbf{0})$$$$\mathbf{f}_i = \boldsymbol{\Lambda}^T \mathbf{R}^{-1} \mathbf{z}_i \quad \blacksquare$$

## **Part B: Parametric Factor Partition Engine & UI Architecture**

In [8]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# =====================================================================
# SIMULATED METRIC GENERATOR (Hardware Sensor Position Framework)
# =====================================================================
def generate_empirical_telemetry(n_snapshots=1500, m_sensors=6):
    """Generates multi-channel engineering snapshot observations."""
    np.random.seed(42)
    # Define primary underlying latent physical factors
    f1 = np.sin(np.linspace(0, 40, n_snapshots))
    f2 = np.sign(np.cos(np.linspace(0, 20, n_snapshots)))

    # Map factors onto individual channels with an explicit unique noise floor
    data = {
        "Vibration_Ch_A": 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.25, n_snapshots),
        "Vibration_Ch_B": 0.75 * f1 + 0.05 * f2 + np.random.normal(0, 0.35, n_snapshots),
        "Thermal_Ch_A":   0.10 * f1 + 0.85 * f2 + np.random.normal(0, 0.15, n_snapshots),
        "Thermal_Ch_B":   0.05 * f1 + 0.80 * f2 + np.random.normal(0, 0.20, n_snapshots),
        "Acoustic_Floor": 0.40 * f1 + 0.40 * f2 + np.random.normal(0, 0.45, n_snapshots),
        "Static_Channel": np.ones(n_snapshots) * 5.0 # Zero-variance diagnostic channel
    }
    return pd.DataFrame(data)

# =====================================================================
# 1. & 2. PARAMETRIC ENGINE & CONSTRAINT VALIDATION LAYERS
# =====================================================================
def execute_factor_analysis_pipeline(df: pd.DataFrame, k_factors: int = 2) -> dict:
    """
    Executes a structured Factor Analysis pipeline with numeric guardrails,
    Varimax matrix rotation transformations, and state tracking.
    """
    # 1. Z-Score Mapping Layer with specific zero-variance boundary guards
    numeric_df = df.select_dtypes(include=[np.number])
    feature_names = list(numeric_df.columns)

    n, m = numeric_df.shape
    X = numeric_df.to_numpy()

    means = np.mean(X, axis=0)
    stds = np.std(X, axis=0, ddof=1)

    # Explicit numerical guardrail capping zero-variance channels to 10^-15
    stds = np.where(stds == 0, 1e-15, stds)
    Z = (X - means) / stds

    # 2. Dimensionality Constraint Layer validation tests
    if n <= m:
        raise ValueError(
            f"Operational Constraint Error: Snapshot count n ({n}) must be strictly "
            f"greater than isolated feature count m ({m})."
        )
    if k_factors >= m:
        raise ValueError(
            f"Latent Dimensionality Constraint Error: Requested target factor dimension k ({k_factors}) "
            f"must be strictly less than the available feature channel space m ({m})."
        )

    # 3. Information Decomposition Layer (SVD / Maximum Likelihood Integration)
    # Fit Factor Analysis with an orthogonal Varimax rotation matrix (T)
    fa = FactorAnalysis(n_components=k_factors, rotation='varimax', random_state=42)
    fa.fit(Z)

    # Extract structural rotated weight vectors (Lambda)
    lambda_rotated = fa.components_.T  # Shape: (m, k)

    # Compute Communality (h²) and Unique Noise Variance (phi²) vectors
    communality = np.sum(lambda_rotated**2, axis=1)
    uniqueness = fa.noise_variance_

    # 4. Mathematical State Reconstruction via Thomson's MMSE Regression Method
    # Formula: F = Z * R^-1 * Lambda_rotated
    R = np.corrcoef(Z.T)
    R_inv = np.linalg.inv(R)
    score_weights = R_inv @ lambda_rotated
    F = Z @ score_weights

    # 5. Performance Diagnostic Telemetry Allocation (ddof=1)
    empirical_variance_F = np.var(F, axis=0, ddof=1)

    avg_communality_pct = np.mean(communality) * 100
    avg_uniqueness_pct = np.mean(uniqueness) * 100

    # Format out clean system logging tracking telemetry metrics
    print("=" * 70)
    print("             DIAGNOSTIC PIPELINE STATE MATRIX SUMMARY             ")
    print("=" * 70)
    print(f"Average System Communality %: {avg_communality_pct:.4f}%")
    print(f"Average System Uniqueness %:  {avg_uniqueness_pct:.4f}%")
    print("=" * 70)

    return {
        "Feature_Names": feature_names,
        "Lambda_Rotated": lambda_rotated,
        "Communality": communality,
        "Uniqueness": uniqueness,
        "Latent_Variances": empirical_variance_F,
        "k_factors": k_factors
    }

# =====================================================================
# PRODUCTION-GRADE 2 × 2 DIAGNOSTIC VISUALIZATION CANVAS
# =====================================================================
def draw_diagnostics_matrix_dashboard(results: dict):
    """
    Renders an asymmetric 2x2 layout canvas meeting strict geometric spatial constraints.
    """
    features = results["Feature_Names"]
    factors = [f"Factor {i+1}" for i in range(results["k_factors"])]

    # Instantiate Subplot with explicit horizontal_spacing and vertical_spacing configurations
    fig = make_subplots(
        rows=2, cols=2,
        horizontal_spacing=0.24,
        vertical_spacing=0.28,
        subplot_titles=(
            "Structural Loadings Heatmap",
            "Variance Partitioning Profile",
            "Sensor Noise Floor Line",
            "Latent Empirical Variance"
        )
    )

    # -----------------------------------------------------------------
    # Subplot (1,1): Structural Loadings Matrix Heatmap
    # -----------------------------------------------------------------
    abs_loadings = np.abs(results["Lambda_Rotated"])

    fig.add_trace(
        go.Heatmap(
            z=abs_loadings,
            x=factors,
            y=features,
            colorscale="YlOrRd",
            showscale=True,
            colorbar=dict(
                x=-0.15,
                len=0.38,
                y=0.78,
                yanchor="middle",
                xanchor="right",
                title="Loading Map"
            ),
            name="Structural Loading Magnitude"
        ),
        row=1, col=1
    )

    # -----------------------------------------------------------------
    # Subplot (1,2): Variance Partitioning Profile (Horizontal Stacked Bar)
    # -----------------------------------------------------------------
    h2_pct = results["Communality"] * 100
    phi2_pct = results["Uniqueness"] * 100

    # Base Layer: Shared Communality (h²)
    fig.add_trace(
        go.Bar(
            y=features,
            x=h2_pct,
            orientation='h',
            marker=dict(color='#1f77b4'),
            name='Shared Communality (h²)',
            legendgroup='global_legend'
        ),
        row=1, col=2
    )

    # Stacked Layer: Isolated Uniqueness (phi²)
    fig.add_trace(
        go.Bar(
            y=features,
            x=phi2_pct,
            orientation='h',
            marker=dict(color='#ff7f0e'),
            name='Isolated Uniqueness (φ²)',
            legendgroup='global_legend'
        ),
        row=1, col=2
    )

    # -----------------------------------------------------------------
    # Subplot (2,1): Sensor Noise Floor Line Profile
    # -----------------------------------------------------------------
    fig.add_trace(
        go.Scatter(
            x=features,
            y=results["Uniqueness"],
            mode='lines+markers',
            line=dict(color='#d62728', width=2, dash='dashdot'),
            marker=dict(symbol='x', size=8),
            name='Uniqueness Path (φ²)',
            legendgroup='global_legend'
        ),
        row=2, col=1
    )

    # -----------------------------------------------------------------
    # Subplot (2,2): Latent Empirical Variance
    # -----------------------------------------------------------------
    fig.add_trace(
        go.Bar(
            x=factors,
            y=results["Latent_Variances"],
            marker=dict(
                color='#2ca02c',
                line=dict(color='black', width=0.5)
            ),
            name='Empirical Factor Score Variance',
            legendgroup='global_legend'
        ),
        row=2, col=2
    )

    # =====================================================================
    # LAYOUT STRATIFICATION AND GEOMETRIC GEOMETRY ENFORCEMENT
    # =====================================================================
    fig.update_layout(
        template="plotly_white",
        width=1250,
        height=750,
        barmode='stack',
        # Centered horizontal legend located directly above the subplots grid matrix
        legend=dict(
            orientation="h",
            x=0.5,
            y=1.02,
            xanchor="center",
            yanchor="bottom"
        ),
        # Strict explicit canvas margin padding structures
        margin=dict(t=150, b=60, l=140, r=80)
    )

    # Apply label updates and rotation metrics to Subplots
    fig.update_xaxes(title_text="Latent Factors", row=1, col=1)
    fig.update_xaxes(title_text="Allocation Profile (%)", range=[0, 100], row=1, col=2)
    fig.update_xaxes(title_text="Hardware Tracking Positions", tickangle=25, row=2, col=1)
    fig.update_xaxes(title_text="Factor Space Vector", row=2, col=2)

    fig.update_yaxes(title_text="Isolated Sensor Channels", row=1, col=1)
    fig.update_yaxes(title_text="Uniqueness Metric Floor Value", row=2, col=1)
    fig.update_yaxes(title_text="Calculated Scale Variance", row=2, col=2)

    # Output presentation render
    fig.show()

# =====================================================================
# PIPELINE EXECUTION
# =====================================================================
# Step 1: Generate Raw Multi-Channel Telemetry Snapshot
df_empirical_data = generate_empirical_telemetry(n_snapshots=1500, m_sensors=6)

# Step 2: Compute Mathematical Space Parameters
pipeline_outputs = execute_factor_analysis_pipeline(df_empirical_data, k_factors=2)

# Step 3: Draw 2x2 Dashboard Visualization Grid
draw_diagnostics_matrix_dashboard(pipeline_outputs)

/usr/local/lib/python3.12/dist-packages/numpy/lib/_function_base_impl.py:2922: RuntimeWarning:

invalid value encountered in divide

/usr/local/lib/python3.12/dist-packages/numpy/lib/_function_base_impl.py:2923: RuntimeWarning:

invalid value encountered in divide



             DIAGNOSTIC PIPELINE STATE MATRIX SUMMARY             
Average System Communality %: 67.2134%
Average System Uniqueness %:  16.0645%


# **Subspace Diagnostics & Feature De-correlation**

## **Part A: Develop a Feature Analysis Dashboard**

In [11]:
import numpy as np
import pandas as pd
from sklearn.decomposition import FactorAnalysis
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# =====================================================================
# NOMINAL SYSTEM Telemetry GENERATOR (4-Channel Sensor Array)
# =====================================================================
def generate_nominal_shm_data(n_snapshots=2500):
    """
    Generates a high-dimensional, collinear baseline dataset representing
    a 4-channel structural health monitoring sensor grid.
    """
    np.random.seed(101)
    # Underlying common physical driving forces (e.g., thermal cycles, structural vibrations)
    global_force_1 = np.sin(np.linspace(0, 50, n_snapshots))
    global_force_2 = np.cos(np.linspace(0, 30, n_snapshots))

    # Synthesize highly collinear channel readouts with a localized sensor noise floor
    data = {
        "Sensor_1": 0.85 * global_force_1 + 0.10 * global_force_2 + np.random.normal(0, 0.15, n_snapshots),
        "Sensor_2": 0.80 * global_force_1 + 0.20 * global_force_2 + np.random.normal(0, 0.20, n_snapshots),
        "Sensor_3": 0.15 * global_force_1 + 0.90 * global_force_2 + np.random.normal(0, 0.10, n_snapshots),
        "Sensor_4": 0.05 * global_force_1 + 0.05 * global_force_2 + np.random.normal(0, 0.05, n_snapshots) # Lower variance node
    }
    return pd.DataFrame(data)

# =====================================================================
# COMPONENT 1: PCA OPTIMIZATION SUITE ANALYSIS ENGINE
# =====================================================================
def run_pca_optimization_suite(df: pd.DataFrame):
    """
    Computes full non-parametric PCA metrics, variance profiles,
    and multi-subspace Hotelling T2 / Q statistic decay.
    """
    X = df.select_dtypes(include=[np.number]).to_numpy()
    n, m = X.shape

    # Standardize data to zero-mean and unit variance with a numerical safety floor
    means = np.mean(X, axis=0)
    stds = np.std(X, axis=0, ddof=1)
    stds = np.where(stds == 0, 1e-15, stds)
    Z = (X - means) / stds

    # Spectral decomposition of the unbiased sample covariance matrix
    S = (Z.T @ Z) / (n - 1)
    eigenvalues, eigenvectors = np.linalg.eigh(S)

    # Sort components in strictly descending order
    idx = np.argsort(eigenvalues)[::-1]
    eigenvalues = eigenvalues[idx]
    eigenvectors = eigenvectors[:, idx]

    # Compute score vectors array (Z projected onto orthogonal matrix P)
    Z_projected = Z @ eigenvectors

    # Pre-allocate array slots for subspace diagnostics tracking vs. k cutoffs
    k_range = list(range(1, m + 1))
    mean_t2_scores = []
    mean_q_scores = []

    for k in k_range:
        # Hotelling's T2: Mahalanobis distance metric calculation
        Z_k = Z_projected[:, :k]
        lambdas_k = eigenvalues[:k]
        t2_samples = np.sum((Z_k ** 2) / lambdas_k, axis=1)
        mean_t2_scores.append(np.mean(t2_samples))

        # Residual Q Statistic (SPE): Squared Euclidean norm of remainder space
        Z_residual = Z_projected[:, k:]
        q_samples = np.sum(Z_residual ** 2, axis=1)
        mean_q_scores.append(np.mean(q_samples))

    # Percent and cumulative variance tracking computations
    total_var = np.sum(eigenvalues)
    marginal_variance_pct = (eigenvalues / total_var) * 100
    cumulative_variance_pct = np.cumsum(marginal_variance_pct)
    residual_variance_pct = 100 - cumulative_variance_pct

    return {
        "Channels": list(df.select_dtypes(include=[np.number]).columns),
        "Eigenvectors": eigenvectors,
        "Eigenvalues": eigenvalues,
        "Marginal_Var_Pct": marginal_variance_pct,
        "Cumulative_Var_Pct": cumulative_variance_pct,
        "Residual_Var_Pct": residual_variance_pct,
        "K_Range": k_range,
        "Mean_T2": mean_t2_scores,
        "Mean_Q": mean_q_scores
    }

# =====================================================================
# COMPONENT 2: FACTOR ANALYSIS LATENT SUITE ANALYSIS ENGINE
# =====================================================================
def run_fa_latent_suite(df: pd.DataFrame, k_factors=2):
    """
    Fits generative Factor Analysis using Varimax rotation, performing
    Thomson's MMSE regression for factor scores.
    """
    X = df.select_dtypes(include=[np.number]).to_numpy()
    n, m = X.shape

    # Safe standardization mapping layer
    means = np.mean(X, axis=0)
    stds = np.std(X, axis=0, ddof=1)
    stds = np.where(stds == 0, 1e-15, stds)
    Z = (X - means) / stds

    # Run scikit-learn SVD / Maximum Likelihood engine with Varimax structural rotation
    fa = FactorAnalysis(n_components=k_factors, rotation='varimax', random_state=42)
    fa.fit(Z)

    lambda_rotated = fa.components_.T # Extracted matrix loads
    communality = np.sum(lambda_rotated**2, axis=1)
    uniqueness = fa.noise_variance_

    # Thomson’s Regression Estimator: F = Z * R^-1 * Lambda_rotated
    R = np.corrcoef(Z.T)
    R_inv = np.linalg.inv(R)
    F = Z @ R_inv @ lambda_rotated

    # Empirical post-rotation factor scores variance (ddof=1)
    factor_variances = np.var(F, axis=0, ddof=1)

    return {
        "Channels": list(df.select_dtypes(include=[np.number]).columns),
        "Lambda_Rotated": lambda_rotated,
        "Communality": communality,
        "Uniqueness": uniqueness,
        "Factor_Variances": factor_variances,
        "k_factors": k_factors
    }

# =====================================================================
# UI LAYOUT BUILDER 1: PCA OPTIMIZATION SUITE (2x3 GRID)
# =====================================================================
def render_ui_layout_pca(pca_metrics):
    """Generates the unified 6-panel PCA optimization report canvas."""
    features = pca_metrics["Channels"]
    pc_labels = [f"PC_{i+1}" for i in range(len(features))]

    fig = make_subplots(
        rows=2, cols=3,
        horizontal_spacing=0.15,
        vertical_spacing=0.25,
        subplot_titles=(
            "Feature Loadings Matrix Heatmap |P|",
            "Absolute Eigenvalues (Variance Magnitude)",
            "Captured Variance: Marginal vs. Cumulative",
            "Residual Space (Unexplained Variance %)",
            "Sample Mean Hotelling's T² vs. Subspace Size k",
            "Sample Mean Q Statistic (SPE) vs. Cutoff k"
        )
    )

    # Panel 1,1: Loadings Heatmap
    fig.add_trace(go.Heatmap(
        z=np.abs(pca_metrics["Eigenvectors"]), x=pc_labels, y=features,
        colorscale="Viridis", showscale=False), row=1, col=1)

    # Panel 1,2: Eigenvalues Bar
    fig.add_trace(go.Bar(
        x=pc_labels, y=pca_metrics["Eigenvalues"],
        marker=dict(color="#2ca02c")), row=1, col=2)

    # Panel 1,3: Marginal vs Cumulative Energy
    fig.add_trace(go.Bar(
        x=pc_labels, y=pca_metrics["Marginal_Var_Pct"],
        name="Marginal Var %", marker=dict(color="#1f77b4")), row=1, col=3)
    fig.add_trace(go.Scatter(
        x=pc_labels, y=pca_metrics["Cumulative_Var_Pct"],
        name="Cumulative Var %", line=dict(color="#d62728", dash="dash", width=2.5)), row=1, col=3)

    # Panel 2,1: Residual Information Leakage
    fig.add_trace(go.Bar(
        x=pc_labels, y=pca_metrics["Residual_Var_Pct"],
        marker=dict(color="#bcbd22")), row=2, col=1)

    # Panel 2,2: Hotelling T² Metric Evolution Curve
    fig.add_trace(go.Scatter(
        x=pca_metrics["K_Range"], y=pca_metrics["Mean_T2"],
        mode="lines+markers", line=dict(color="#9467bd", width=2)), row=2, col=2)

    # Panel 2,3: SPE Q Metric Residual Decay Path
    fig.add_trace(go.Scatter(
        x=pca_metrics["K_Range"], y=pca_metrics["Mean_Q"],
        mode="lines+markers", line=dict(color="#e377c2", width=2)), row=2, col=3)

    fig.update_layout(
        template="plotly_white", width=1350, height=800, showlegend=False,
        margin=dict(t=80, b=60, l=80, r=40)
    )
    fig.update_xaxes(title_text="Subspace Cutoff (k)", row=2, col=2)
    fig.update_xaxes(title_text="Truncation Cutoff (k)", row=2, col=3)
    fig.show()

# =====================================================================
# UI LAYOUT BUILDER 2: FA LATENT SUBSPACE SUITE (2x2 GRID)
# =====================================================================
def render_ui_layout_fa(fa_metrics):
    """Generates the unified balanced 4-panel FA latent diagnostics suite."""
    features = fa_metrics["Channels"]
    factor_labels = [f"Factor_{i+1}" for i in range(fa_metrics["k_factors"])]

    fig = make_subplots(
        rows=2, cols=2,
        horizontal_spacing=0.24,
        vertical_spacing=0.28,
        subplot_titles=(
            "Structural Loadings Matrix Heatmap |λ|",
            "Variance Partitioning Profile (h² vs. φ²)",
            "Sensor Uniqueness Noise Floor Profile",
            "Latent Factor Scores Empirical Variance"
        )
    )

    # Panel 1,1: Heatmap of Absolute Loads
    fig.add_trace(go.Heatmap(
        z=np.abs(fa_metrics["Lambda_Rotated"]), x=factor_labels, y=features,
        colorscale="YlOrRd", showscale=True,
        colorbar=dict(x=-0.15, len=0.38, y=0.78, yanchor="middle", xanchor="right")), row=1, col=1)

    # Panel 1,2: Stacked Horizontal Bar Allocation Chart (h² vs phi²)
    fig.add_trace(go.Bar(
        y=features, x=fa_metrics["Communality"] * 100,
        orientation='h', marker=dict(color='#1f77b4'), name="Communality (h²)"), row=1, col=2)
    fig.add_trace(go.Bar(
        y=features, x=fa_metrics["Uniqueness"] * 100,
        orientation='h', marker=dict(color='#ff7f0e'), name="Uniqueness (φ²)"), row=1, col=2)

    # Panel 2,1: Red Line Sensor Noise Floors
    fig.add_trace(go.Scatter(
        x=features, y=fa_metrics["Uniqueness"],
        mode='lines+markers', line=dict(color='#d62728', width=2, dash='dashdot'),
        marker=dict(symbol='x', size=8)), row=2, col=1)

    # Panel 2,2: Empirical Factor Post-Rotation Variances
    fig.add_trace(go.Bar(
        x=factor_labels, y=fa_metrics["Factor_Variances"],
        marker=dict(color='#2ca02c', line=dict(color='black', width=0.5))), row=2, col=2)

    fig.update_layout(
        template="plotly_white", width=1250, height=750, barmode='stack',
        legend=dict(orientation="h", x=0.5, y=1.02, xanchor="center", yanchor="bottom"),
        margin=dict(t=150, b=60, l=140, r=80)
    )
    fig.update_xaxes(range=[0, 100], row=1, col=2)
    fig.update_xaxes(tickangle=25, row=2, col=1)
    fig.show()

# =====================================================================
# OPERATIONAL PIPELINE EXECUTION ENTRYPOINT
# =====================================================================
# Step 1: Generate multi-channel physical telemetry dataset
df_shm_data = generate_nominal_shm_data(n_snapshots=2500)

# Step 2: Extract statistical features via non-parametric geometric mapping (PCA)
pca_results = run_pca_optimization_suite(df_shm_data)
render_ui_layout_pca(pca_results)

# Step 3: Extract features via parametric noise stratification mapping (FA)
fa_results = run_fa_latent_suite(df_shm_data, k_factors=2)
render_ui_layout_fa(fa_results)

## **Part B: Analysis**

In [15]:
# =====================================================================
# EXTRACTION AND VERIFICATION OF PART B: VARIMAX SIMPLE STRUCTURE
# =====================================================================

# 1. Reuse the FA pipeline outputs from the previous execution
# The 'engine' object and 'compute_empirical_fa' method are not defined in this context.
# We will use the 'pipeline_outputs' dictionary already computed in the notebook.
fa_outputs = pipeline_outputs

# 2. Isolate the Varimax-rotated loading matrix Λ*
# The key 'factor_loadings_lambda' is not present in pipeline_outputs, use 'Lambda_Rotated' instead.
lambda_rotated = fa_outputs["Lambda_Rotated"]
# The key 'sensors' is not present in pipeline_outputs, use 'Feature_Names' instead.
sensor_names = fa_outputs["Feature_Names"]

# 3. Create a clean DataFrame to inspect the "Simple Structure"
df_loadings = pd.DataFrame(
    data=lambda_rotated,
    index=sensor_names,
    columns=["Factor 1 (Primary Process)", "Factor 2 (Secondary Process)"]
)

# 4. Calculate column-wise variance of the squared loadings to quantify the Varimax optimization
# (Varimax maximizes this variance to drive loadings toward binary ~0 or ~1 states)
squared_loadings_variance = np.var(lambda_rotated**2, axis=0)

# =====================================================================
# PRINT DIAGNOSTIC RESULTS TO CONSOLE
# =====================================================================
print("\n" + "="*65)
print("        PART B VERIFICATION: VARIMAX LOADING MATRIX (Λ*)        ")
print("="*65)
print(df_loadings.round(4))
print("-"*65)
print("Varimax Optimization Metric (Variance of Squared Loadings):")
for idx, val in enumerate(squared_loadings_variance):
    print(f"  Factor {idx+1}: {val:.5f} (High variance indicates clean, simple binary groupings)")
print("="*65)

# 5. Programmatic interpretation wrapper for the plant operator
print("\n[Operator Automated Diagnostics System Announcement]:")
for sensor in sensor_names:
    loadings = df_loadings.loc[sensor].abs()
    if loadings.max() < 0.2:
        print(f"  -> {sensor}: Isolated as dead/faulty channel (Captured cleanly by Uniqueness φ²).")
    else:
        primary_factor = loadings.idxmax()
        print(f"  -> {sensor} dynamically assigned to: {primary_factor}")
print("="*65)


        PART B VERIFICATION: VARIMAX LOADING MATRIX (Λ*)        
                Factor 1 (Primary Process)  Factor 2 (Secondary Process)
Vibration_Ch_A                     -0.2303                       -0.8986
Vibration_Ch_B                     -0.0555                       -0.8410
Thermal_Ch_A                       -0.9758                       -0.1359
Thermal_Ch_B                       -0.9623                       -0.1137
Acoustic_Floor                     -0.5715                       -0.4750
Static_Channel                      0.0000                        0.0000
-----------------------------------------------------------------
Varimax Optimization Metric (Variance of Squared Loadings):
  Factor 1: 0.17028 (High variance indicates clean, simple binary groupings)
  Factor 2: 0.11343 (High variance indicates clean, simple binary groupings)

[Operator Automated Diagnostics System Announcement]:
  -> Vibration_Ch_A dynamically assigned to: Factor 2 (Secondary Process)
  -> Vibration

### Module 1: The High-Variance Noise Illusion (PCA vs. FA)

#### 1. Physical Nature of Sensor 4 Telemetry

A **Uniqueness value ($\varphi^2$) exceeding $98\%$** indicates that almost the entirety of the variance observed in `Sensor_4` is completely decoupled from the underlying shared structural dynamics ($f_1$ and $f_2$) of the asset grid.

* **Physical Reality:** In an industrial plant setting, this is the classic mathematical signature of an internal instrument fault. It proves that `Sensor_4` is no longer operating as a functional observation node for structural mechanics; instead, it has degenerated into an isolated source of pure localized entropy.
* **Root Cause Diagnostics:** This behavior typically points to a hardware-level instrumentation breakdown, such as a broken transducer shield, a loose coaxial connector, a localized ground loop, or severe electromagnetic interference affecting that specific sensor's analog-to-digital converter (ADC).

---

#### 2. The PCA Eigenvalue "Trick" Mechanics

The non-parametric optimization engine of Principal Component Analysis identifies orthogonal directions that maximize total projected variance by performing an unscaled spectral decomposition of the sample covariance matrix $\mathbf{S}$. Because PCA calculates global variance without discriminating between its structural or stochastic origins, its mathematical scope is bounded by the total trace of the system:

$$\text{tr}(\mathbf{S}) = \sigma_{\text{Sensor\_1}}^2 + \sigma_{\text{Sensor\_2}}^2 + \sigma_{\text{Sensor\_3}}^2 + \sigma_{\text{Sensor\_4}}^2$$

In your simulation, `Sensor_4` has a massive localized noise contamination scale ($\sigma^2 \approx 1.40^2 \approx 1.96$), causing its individual variance magnitude to swell to $\sigma^2 \approx 2.0$. This completely dominates the scale of the healthy, tightly coupled signals (`Sensor_1` to `Sensor_3`).

Because PCA treats all variance uniformly, the geometric optimization routine rotates its top principal axes directly toward the high-amplitude, random noise coordinate axis of `Sensor_4` to maximize variance capture. This inflates the top eigenvalues ($\lambda_1$ and $\lambda_2$), leading an engineer to believe that these components represent highly energetic, meaningful structural modes, when they are actually capturing localized instrumentation noise.

---

#### 3. Automated Anomaly Detection Risks

Relying strictly on an unpartitioned PCA engine under these conditions creates a major vulnerability in automated monitoring and plant protection loops:

* **High Rate of False Alarms (Type I Errors):** Any trivial fluctuation, electrical spike, or minor baseline drift on the faulty `Sensor_4` channel will project directly onto the primary principal subspace, blowing past control limits and triggering immediate false plant shutdowns.
* **Missed True Defects / Blind Spots (Type II Errors):** Because the primary PCA subspace is heavily distorted by the noise axis of `Sensor_4`, actual, low-amplitude structural deviations (e.g., micro-cracking, structural shifting, or mechanical looseness) occurring in the healthy, tightly coupled sensors (`Sensor_1` to `Sensor_3`) will be masked or squeezed out of the top components, allowing real structural failures to progress undetected.

---

### Module 2: Latent Spaces & Structural Interpretability

#### 1. How Varimax Rotation Relaxes the PCA Hierarchy to Create a "Simple Structure"

Traditional PCA forces its principal axes to follow a strict mathematical hierarchy ($\lambda_1 > \lambda_2 > \dots > \lambda_m$) where every component tries to sweep up as much mixed variance as possible across all sensors. This results in a "dense loading structure" where every sensor has moderate-to-high coefficients across all components, blending separate underlying physical phenomena into a single abstract mathematical axis.

Factor Analysis relaxes this strict variance-maximization hierarchy. It introduces an orthogonal rotation matrix $\mathbf{T}$ (where $\mathbf{T}\mathbf{T}^T = \mathbf{I}$) to transform the raw factor loading matrix into a **Varimax simple structure**:

$$\boldsymbol{\Lambda}^* = \boldsymbol{\Lambda}\mathbf{T}$$

The Varimax criterion maximizes the variance of the *squared loadings* within the columns of the loading matrix:

$$V = \frac{1}{m} \sum_{l=1}^k \sum_{j=1}^m \left( \lambda_{jl}^{*2} - \bar{\lambda}_l^2 \right)^2$$

This mathematically drives the loadings toward binary states: either very high (near $\pm 1$) or near zero. It isolates distinct physical sub-phenomena onto separate latent axes, grouping correlated channels together while stripping away cross-channel contamination. Your specific configuration isolates the structural space into clean groups:

* **Factor 1 Axis:** Tracks the primary process structural loads ($f_1$), showing heavy loadings exclusively for `Sensor_1` and `Sensor_2`.
* **Factor 2 Axis:** Tracks the secondary operational mode ($f_2$), mapping exclusively to `Sensor_3`.

---

#### 2. Practical Troubleshooting Utility: Plant Operator Perspective

From a plant operator’s standpoint, a raw PCA eigenvector matrix is incredibly difficult to interpret during a crisis because a fault signal is smeared across all principal components simultaneously. If a physical failure or degradation occurs in a specific subsystem, its signature will be mixed across `PC 1` and `PC 2`, making it impossible to deduce the root cause from a diagnostic screen.

In contrast, the rotated FA loading matrix provides clear, isolated groupings that align with the actual physical setup of the asset:

* If **Factor 1** experiences an anomalous shift, the operator knows instantly to investigate the physical assets tied to `Sensor_1` and `Sensor_2` (e.g., primary structural load, main shaft torsion, or structural bending stress).
* If **Factor 2** shifts, attention can be directed exclusively to the subsystem monitored by `Sensor_3` (e.g., secondary operational mode, thermal expansion, or auxiliary fluid loops).
* If `Sensor_4` spikes, the FA framework recognizes it has a high uniqueness value, suppressing it at the factor level so it does not trigger a global alarm, instead generating a low-priority instrumentation maintenance ticket.

---

### Module 3: Truncation Boundaries & Subspace Analytics

#### 1. Exact Behavior of the Mean $Q$ Statistic Curve

The Mean $Q$ Statistic (Squared Prediction Error, or SPE) calculates the residual error energy cast into the discarded residual subspace after projecting observations onto a truncated $k$-dimensional subspace:

$$Q_i = \mathbf{x}_i \left( \mathbf{I} - \hat{\mathbf{P}}_k \hat{\mathbf{P}}_k^T \right) \mathbf{x}_i^T$$

* **Transition from $k=1$ to $k=2$:** The Mean $Q$ Statistic curve drops sharply. This occurs because moving from a 1D to a 2D subspace integrates the second true physical driving force ($f_2$), recovering a substantial amount of correlated system energy and removing it from the residual space.
* **Transition from $k=2$ to $k=3$:** The curve flattens out completely, forming a distinct **"flat elbow."** Because there are no more true underlying physical factors present in the data-generation equations, adding a third dimension yields practically zero structural energy recovery.

---

#### 2. Identifying True Physical Dimensionality ($k=2$)

The sharp drop followed by a distinct "flat elbow" in the $Q$ statistic curve identifies **$k=2$** as the true hidden physical dimensionality of this asset. This pattern indicates that exactly two independent physical phenomena ($f_1$ and $f_2$) govern the structural behavior of the asset. Any dimensions extracted beyond this point do not capture structural correlations; they simply partition background noise.

---

#### 3. Risk of Mis-specifying Truncation Cutoff to $k=3$

If you incorrectly choose a truncation cutoff of $k=3$ for your monitoring system, you force the model to treat random noise as a clean component. Specifically, this forces the unmodelable, high-amplitude white noise floor of the broken `Sensor_4` out of the residual tracking space and directly into the primary principal subspace. This introduces background noise into the baseline model, which degrades the sensitivity of Hotelling’s $T^2$ to actual structural changes.

---

### Module 4: Asset Fleet Predictive Maintenance Strategy

#### 1. Operational Strategy Trade-offs

| Strategic Matrix | The PCA Strategy ($T^2$ & $Q$ Monitoring) | The FA Strategy (Direct Latent Factor Tracking) |
| --- | --- | --- |
| **Operational Methodology** | Projects raw observations onto global orthogonal axes. Tracks distance within the subspace ($T^2$) and residual leakage ($Q$). | Partitions shared variance from unique variance. Maps variables directly onto rotated latent components ($F$). |
| **Sensitivity to Localized Noise** | **High.** High-amplitude localized noise distorts the primary subspace alignment and inflates top eigenvalues. | **Low.** Captures unique localized sensor noise within the diagonal matrix $\boldsymbol{\Psi}$, leaving latent tracking unaffected. |
| **Physical Interpretability** | **Low.** Mixes multiple physical effects across components due to the strict variance-hierarchy constraint. | **High.** Aligns separate physical mechanisms with distinct latent factors using Varimax simple structure. |

---

#### 2. Fleet Robustness Winner: The FA Strategy

The **Factor Analysis Strategy** is significantly more robust against a single sensor losing calibration or experiencing an electrical short.

##### Justification via Uniqueness Noise Floor ($\varphi^2$) Metrics:

Factor Analysis explicitly isolates localized sensor failures through its fundamental covariance structuring:

$$\mathbf{R} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$$

When `Sensor_4` suffers an extreme instrument fault, its massive localized variance is cleanly captured by the $4\text{th}$ diagonal entry of the unique variance matrix $\boldsymbol{\Psi}$ ($\varphi_4^2 > 98\%$). Because this localized noise is partitioned into $\boldsymbol{\Psi}$, it is completely prevented from bleeding into the factor loading matrix $\boldsymbol{\Lambda}$.

When estimating latent states via Thomson’s Minimum Mean Squared Error (MMSE) regression method:

$$\mathbf{F} = \mathbf{Z} \mathbf{R}^{-1} \boldsymbol{\Lambda} = \mathbf{Z} \left( \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi} \right)^{-1} \boldsymbol{\Lambda}$$

Applying Woodbury's identity transforms this estimator into:

$$\mathbf{F} = \mathbf{Z} \boldsymbol{\Psi}^{-1} \boldsymbol{\Lambda} \left( \mathbf{I} + \boldsymbol{\Lambda}^T \boldsymbol{\Psi}^{-1} \boldsymbol{\Lambda} \right)^{-1}$$

Because the standardized data matrix $\mathbf{Z}$ is directly multiplied by the inverse of the uniqueness noise floor matrix ($\boldsymbol{\Psi}^{-1}$), the faulty channel is scaled by $\frac{1}{\varphi_4^2} \approx \frac{1}{0.98}$. Consequently, its mathematical contribution to the calculated factor scores drops significantly compared to the healthy channels.

This inverse-variance scaling enables the FA engine to automatically down-weight the corrupted data streams from `Sensor_4`. As a result, the tracked latent factor scores ($F_1$ and $F_2$) remain stable, clean, and unaffected by localized sensor faults. Under a PCA strategy, an identical sensor short would distort the primary projection space, causing false alarms in Hotelling’s $T^2$ and rendering the automated anomaly detection framework unreliable across the fleet.